Support multiple configuration files.

sttk opened the pull request:
This is the PR about #8.
Please review.

phated commented:
Thanks @sttk - I am currently wrapping up some contract work and should have time to review this soon

phated commented on test/fixtures/search_files.js:
```diff
@@ -0,0 +1,803 @@
+var fs = require('fs');
+
+var homedir;
+if (os.homedir) {
```
use https://github.com/sindresorhus/os-homedir

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+
+
+var initCtx = {
+  homedir: os.homedir ? os.homedir() : require('userhome')(),
```
use https://github.com/sindresorhus/os-homedir

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+      return homedir;
+    }
+
+    if (pth[1] === '/') {
```
can you use path.normalize instead of checking these and slicing?

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    if (isWindows() && pth[1] === '\\') {
+      return path.join(homedir, pth.slice(2));
```
can you use path.normalize instead of checking these and slicing?

sttk replied:
This part is for excluding paths like `'~sttk/aaa'`. So these checking and slicing are needed.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return process.platform === 'win32';
+}
+
+function isFunction(v) {
```
I'd be fine with bringing lodash in for these utilities unless @tkellen has an objection

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return (typeof v === 'function');
+}
+
+function isObject(v) {
```
I'd be fine with bringing lodash in for these utilities unless @tkellen has an objection

jonschlinkert replied:
just a thought, adding lodash to do the `typeof` checks will add ~40-50 ms load time to this lib. doesn't seem like much but 100ms is noticeable to users, and it might push the load time into that range. on the other hand it does make the code a bit cleaner. 

maybe consider something like https://www.npmjs.com/package/kind-of instead? it's very fast and battle tested

phated replied:
As noted in some other comments, I'd like to use the _.mapValues method too, instead of implementing our own (eachProp).  I'm pretty sure this functionality is going to add much more than the 100ms you mentioned because it has lots of filesystem traversing to do.

jonschlinkert replied:
make sense

tkellen replied:
I'm +1 to lodash. The NIH stuff in this codebase was originally written pursuant to what @jonschlinkert mentioned (trying to keep load time as fast as possible).

sttk replied:
@tkellen Thank you for your comment. I'll use lodash.

sttk replied:
On my PC, load time of lodash was about 50ms, and total load time of lodash.isobject, .isarray, .isstring, .isfunction, .identity, and .mapvalues was about 10ms.

Should we use only needed function's modules of lodash?

phated replied:
Yeah, that's fine. The main version of lodash ships with modules now so you could try 'lodash/isObject' also

sttk replied:
Load time of 'lodash/isObject' is as short as 'lodash.isobject' and other 'lodash/is*' modules are also same. But 'lodash/mapValues' takes nearly 50ms because it would load many modules via require tree. (lodash.mapvalues takes less than 10ms and it loads only three modules.)

And this isObject function corresponds to lodash.isplainobject.

phated replied:
@sttk thanks for looking into that.  I'm fine with the npm modules (`lodash.mapvalues`, etc).  Please note that isObject is not the same as isPlainObject, lodash actually has 3 methods for objects (isPlainObject, isObject and isObjectLike)

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return ({}.toString.call(v) === '[object Object]');
+}
+
+function isArray(v) {
```
I'd be fine with bringing lodash in for these utilities unless @tkellen has an objection

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return Array.isArray(v);
+}
+
+function isString(v) {
```
I'd be fine with bringing lodash in for these utilities unless @tkellen has an objection

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return (typeof v === 'string');
+}
+
+function doNothing(v) {
```
I'd be fine with bringing lodash in for these utilities unless @tkellen has an objection.  This is `identity` in lodash

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+
+
+function parsePath(pth) {
+  return path.parse ? path.parse(pth) : require('path-parse')(pth);
```
I'm good with always using the ponyfill - we should also ask @jonschlinkert what he thinks between his version (https://github.com/jonschlinkert/parse-filepath) and the path-parse module.  I'd like his to be 1.0 if we use it.

jonschlinkert replied:
thanks for asking. a few thoughts:

- parse-filepath caches so it's theoretically going to be faster if identical paths are parsed twice. probably not much benefit here, but thought I'd mention
- parse-filepath uses native node.js methods for parsing paths whenever possible, so if bug fixes are made in node.js it's going to use those. It also uses `is-absolute` which is used by npm itself
- parse-path uses regex that was in node.js at some point, I'm not sure how often it's updated so it might be just as reliable.

jonschlinkert replied:
also I don't mind bumping to 1.0 if you use parse-filepath. it's time to bump it to 1.0 anyway.

phated replied:
please bump to 1.0.  thanks!

jonschlinkert replied:
done! I updated it this morning. thanks

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+}
+
+function isAbsolutePath(pth) {
+  return path.isAbsolute ? path.isAbsolute(pth) : !!parsePath(pth).root;
```
I'm fine always using the ponyfill.  I think @jonschlinkert has one of those too

jonschlinkert replied:
yeah [is-absolute](https://www.npmjs.com/package/is-absolute) is used by npm (which)

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  if (isObject(exts)) {
+    return Object.keys(exts).sort();
```
would you want to sort by the object order?  I think lodash has that now

sttk replied:
I did it because of consistent order when multiple paths with different extensions were matched.

phated replied:
`.sort()` in JS isn't stable.  I'd prefer either `sortBy` or `orderBy` from lodash

sttk replied:
This part is for getting a property array of `exts` object and each properties is unique. Since being unstable of `.sort` is about comparing equal, there should be no problem in this case.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  }
+
+  if (!findUp) {
+    return eachUnNullElem(glob.sync(pth, globOpt), toFullPath, globOpt.cwd);
```
can we avoid glob?  The Windows bugs are really frustrating

jonschlinkert replied:
is `findFile` just doing what [findup-sync](https://github.com/js-cli/node-findup-sync) does?

sttk replied:
@jonschlinkert Yes. At first I use findup-sync(v0.3.0) but it returns only one path. So I changed a bit of its code because I wanted to get all matched paths when using `searchPaths` property.

If the latest findup-sync could return all matched paths, I would use it...

phated replied:
hmmm, wonder if we should update findup-sync.  @jonschlinkert thoughts?

jonschlinkert replied:
I think findup-sync should always return a single path, there are a couple of libs that might be better for this.

However, I spent some time with the pr yesterday, @sttk did a great job, but it got me wondering if there might be an opportunity to expand on what liftoff does to add support for resolving globally installed modules (like generators for yeoman, assemble or verb). If so, I would probably approach this feature differently. I recently wrote some code for assemble/verb that does what I mentioned with global modules, so if that sounds interesting - and within the scope of liftoff, I'd love to repurpose that code and move it here so we can consolidate efforts (fwiw, to me it seems like it would make a lot of sense). If not, no worries.

tkellen replied:
I'm cautiously :+1: on this. I believe that Liftoff is too complex already, but only because some of the config finding logic needs to become a standalone module. If the functionality you are describing were added as a dependent library that Liftoff consumes, this could be very valuable.

jonschlinkert replied:
> I believe that Liftoff is too complex already

yeah I was hesitant to bring it up for that reason. I'll put something together with that in mind.

> If the functionality you are describing were added as a dependent library that Liftoff consumes, this could be very valuable.

That makes a lot of sense. I have code that does this that need to be generalized a bit more. I'll let you know when I have something worth reviewing. 

Maybe I'll start with pseudo-code or some kind of graphic to show what I'm thinking. At least you'll have something to look at and decide if it makes sense. (in the meantime don't hold up this pr on my account. I'll do this under the assumption that we're just test driving this idea)

phated replied:
@jonschlinkert awesome.  Could you open an issue to track this?   I'd also like you to expand on the global lookup in that issue (I'm not sure what it will be used for).

jonschlinkert replied:
sure. I'll create one as soon as I have a chance. thanks

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+      return pth;
+    }
+
+  } catch (e) {}
```
why is this error thrown away?

sttk replied:
Ah, since this `pth` file always exists, this `lstatSync` does not throw Error. I'll remove this try-catch.

sttk replied:
About this part, `fs.readlpathSync` throws an error if the specified path is a symbolic link to non-existing file. (There is a test case about this.) So try-catch is needed for `fs.realpathSync`.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+};
+
+function normalizeContext(ctx) {
+  ctx = eachProp(ctx || {}, doNothing);
```
Looks like this can be replaced with lodash.mapValues

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+
+function search(opts, ctx) {
+  ctx = normalizeContext(ctx);
+  return eachProp(opts || {}, searchWithOpt);
```
Looks like this can be replaced with lodash.mapValues

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+    if (!opt || isString(opt) || isArray(opt)) {
+      opt = { searchPaths: opt };
+    } else {
+      opt = eachProp(opt, doNothing);
```
Looks like this can be replaced with lodash.mapValues

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  var fn = isAbsolutePath(name) ? fnNotUseDir : fnEachDirArray;
+
+  if (isObject(dirs) && !('path' in dirs)) {
+    return eachProp(dirs, fn);
```
Looks like this can be replaced with lodash.mapValues

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,291 @@
+  return ret;
+}
+
+function eachProp(obj, fn /*, ...*/) {
```
Looks like this can be replaced with lodash.mapValues

phated commented:
@sttk this PR looks really cool but there's a ton of code in here.  I made some cursory comments on ways to reduce the code and allow me to focus on the logic.  I think another thing that would help would be to have tests written out instead of generated from the fixture you created.

tusbar commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
 node_modules
+coverage
+.DS_Store
```
`.DS_Store` should not be in here.

If you want to ignore those system files, [do it globally](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

phated replied:
I'm fine with this here.  I don't do it globally either and I bet a lot of people don't

tusbar commented on .travis.yml:
```diff
@@ -2,7 +2,8 @@ language: node_js
   - "0.12"
-  - "iojs"
+  - "4"
+  - "stable"
```
For consistency, I would rather see `5` here.

phated replied:
I like to keep this as stable (I wish we could reference LTE instead of 4 also).  This is the pattern used in all the gulpjs repos

sttk commented:
@phated @jonschlinkert @tusbar Thank you all for reviewing and advising. I'll modify all of what you pointed out.

@jonschlinkert I will use parse-filepath, is-absolute!

@phated I welcome the further pointed out. And I expect a better way of testing. Thanks.

phated commented:
@sttk awesome, thanks.  Let me know when the updates are done.

sttk commented:
I've finished modifying the codes.

I stopped using glob and `fs.existsSync` is used simply. So case sensitivity of file paths is platform-dependent. (I may tackle this later.)

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,253 @@
+var isAbsolutePath = require('is-absolute');
+var parsePath = require('parse-filepath');
+
+function isWindows() {
```
this can't change, so it should probably just be a variable defined here `var isWindows = process.platform === 'win32'`

sttk replied:
I've modified it.

tkellen commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+var fs = require('fs');
```
this should be `const fs...`

phated replied:
@tkellen is `const` available in node 0.10?  We support that far back

tkellen commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+var path = require('path');
+var oshomedir = require('os-homedir');
+
+var isFunction = require('lodash.isfunction');
```
all of these requires should be `const ... = require('...');`

phated replied:
@tkellen is `const` available in node 0.10?  We support that far back

tkellen replied:
Ya. IIRC the semantics around how it works aren't 100% correct back that far, but it's present all over this codebase already without issue.

sttk replied:
I've modified them.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+
+function search(opts, ctx) {
```
In `buildEnvironment`, this is never called with a context.  Why is it a parameter?

sttk replied:
I forgot to pass `cwd` to this module in buildEnvironment. I'll modify to pass it.
About home directory, I thought user specified homedir but now I think it isn't needed. So I'll remove the codes about it.

phated replied:
Please add an extra test for this if possible since it seems to be a bug and we wouldn't want a regression

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+};
+
+function normalizeContext(ctx) {
+  ctx = mapValues(ctx || {}, doNothing);
```
Is this just a shallow clone?  Could probably get away with `_.assign({}, initCtx, ctx)`

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+function normalizePath(pth, ctx) {
+  if (isFunction(pth)) {
```
Why are functions supported?  I didn't see that in the spec

sttk replied:
I thought to support cases that users specify a function to `ctx.cwd` or `ctx.homedir`.
But this code will be removed with `normalizePath`.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+
+function normalizeHomedir(pth, homedir) {
```
more @jonschlinkert code!  should probably just use https://github.com/jonschlinkert/expand-tilde instead of a custom module here

sttk replied:
OK. I'll use expand-tilde.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+
+  pth = normalizeHomedir(pth, ctx.homedir);
+
+  if (!isAbsolutePath(pth)) {
```
Instead of this code branch + `path.normalize` below, I think you can use `path.resolve(ctx.cwd, pth)`.  If the 2nd argument is absolute, it should just return it.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+
+var initCtx = {
```
can these be used directly in the code if the `ctx` parameter is removed from the `search` function?

sttk replied:
I'll modify to pass only `cwd` instead of the `ctx` parameter from buildEnvironment.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+}
+
+function normalizeExtensions(exts) {
+  if (isFunction(exts)) {
```
What is the point of supporting a function for extensions, especially if it is sync only?  I see @tkellen had it in the original spec so maybe he has a use case in mind.

tkellen replied:
The original spec didn't take into account js-interpret (it didn't exist then).

phated replied:
how does the extension config interact with `interpret`?

tkellen replied:
The extensions format should match interpret. The example code that used a function in the spec was this:
```
      extensions: function () {
        return Object.keys(require.extensions)
      },
```

...which was basically a rudimentary way of trying to detect which extensions were valid.

The extensions format should match js-interpret so we can use the same logic to handle registering loaders.

tkellen replied:
Actually, this entire set of functionality should be a standalone lib that we can use for all extension files.

phated replied:
I didn't think these files triggered an auto require of the loaders.  If that is the case, there's a lot more work that needs to go into this and I think it would increasingly slow it down.

sttk replied:
About applying by each extension, how do you think about my comment in #8.

> And now, I'm considering that searched paths should be returned not like 'searched/file/path.json' but like { extension: '.json', path:'searched/file/path.json' } for applying require functions corresponding to their extensions after.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+
+function search(opts, ctx) {
+  ctx = normalizeContext(ctx);
+  return mapValues(opts || {}, searchWithOpt);
```
This shouldn't need to be defaulted.  lodash handles undefined values by just not executing the callback

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+  return mapValues(opts || {}, searchWithOpt);
+
+  function searchWithOpt(opt, optname) {
+    if (!opt || isString(opt) || isArray(opt)) {
```
how would `opt` ever be falsey?  Also, if it were, wouldn't we want to exit early instead of set it as an object with falsey `searchPaths` option?

sttk replied:
I think cases as follows:
```js
configFiles = {
  '.apprc': null,
  '/aaa/bbb/ccc.ddd': '',                                                       
  '~/ddd.eee': null,
};

phated replied:
Is the `searchPaths` option not mandatory?  Do you only use that when you need to specify `extensions`?  Maybe `extensions` shouldn't be in the top level config.

sttk replied:
I think the `searchPaths` option is not mandatory. 
If `configFiles` option was null, to exit early is also OK because null can be regarded as specifying an empty array. But if an empty string, I think it better to use the only option name as a path.

phated replied:
@sttk I just realized this last night... the keys aren't supposed to be filepaths.  They are supposed to be the keys that are being assigned to, e.g.
```js
configFiles = {
  rcFile: '.apprc',
  deepPath: '/aaa/bbb/ccc.ddd',
  fileInHome: '~/ddd.eee',
};
```

The paths are then expanded and assigned to the keys so the consuming application can just access the filepath by the named key they want to use.

sttk replied:
@phated At the first @tkellen's comment in #8, it seems that `configFiles` keys '.app' and 'appfile' are used as a part of filepaths. Is the behavior when `configFiles` option is set a string or an array different from when the option is set an object? Or isn't it about `searchPath/searchPaths` options?

phated replied:
hmmm, I suppose you are right but maybe we need to rethink that spec a little.  I don't think we need to support separate extensions (they are already passed to Liftoff as `extensions`) and if that is the case, we don't need to support the `searchPaths` key so everything could be under the top level key. e.g.
```js
configFiles = {
  '.app': ['~/config/', './'. './.config/'],
  'ccc': [{ path: '/aaa/bbb/', findup: true }]
}
```

Is there any reason to support the object?  I think arrays would work fine for every use case.  I also think that the files actually need to be loaded and the resulting config objects to be merged together (this could be done by the array index & using lodash.merge).  @tkellen do you have anything thoughts on this?  I think it would make it much simpler.

phated replied:
oh, and the object that contains `path` and `findup` could be expanded to support extensions if absolutely needed (in a future revision).

phated replied:
@sttk sorry about re-envisioning the scope, but this is the first time I'm starting to see it all come together.

sttk replied:
@phated Don't mind me. In addition, we should contain a key for getting only first found path in the object optionally, instead of separation with `searchPath` and `searchPaths`.

phated replied:
How about `{ path: '/some/path', findup: true, exitEarly: true }`?

phated replied:
btw, with this re-envisioning, we would just need to expand strings to `{ path: 'the string', findup: false, exitEarly: false }` and the process everything the same

sttk replied:
> How about { path: '/some/path', findup: true, exitEarly: true }?

Good! And what about extensions?
And about expanded string, `exitEarly` flag would not be needed.

For example, if multiple paths which were different with only extensions were hit, would it be expanded as follows?

```js
(setting)
configFiles = {
  '.app': [
    { path: '/some/path', extension:['rc', '.yml'] },
  ]
}
--
(result)
configFiles = {
  '.app': [
    { path: '/some/path/.apprc', extension:'rc', findup: false },  // <-- Only this if exitEarly is true in setting
    { path: '/some/path/.app.yml', extension:'.yml', findup: false },
  ]
}
```

sttk replied:
@phated Ah! I might misunderstand expanded strings as result strings. I'm sorry if so.

sttk replied:
(I'm sorry for such a long comment.)

Is the spec right as follows? 
(In such form, as you supposed above, it wouldn't be necessary to use `configFiles` option key as a part of filepath. In addition, I'd like to support user specified key which isn't used here but is inherited for being used after.)

```js
(setting)
configFiles = {
  rcFile: '.apprc',
  configs: ['~/config/.app', './.config/app.conf'],
  aaa: { path: 'path/to/config/file' },
  bbb: { path: ['path/to/file', '/root/path/to/file'], findup: true, exitEarly: true  },
  ccc: [
    { path: ['/root/path/to/file', 'path/to/file'], findup: true, exitEarly: true },
    { path: ['ddd/eee', 'fff/ggg'], exitEarly: true }
  ],
  hhh: { path: 'path/to/file', userSpecKey: 'iii' },
}
--
(expand, and process all in same way)
configFiles = {
  rcFile: [
    { path: '.apprc', findup: false, exitEarly: false },
  ],
  configs: [
    { path: '~/config/.app', findup: false, exitEarly: false },
    { path: './.config/app.conf', findup: false, exitEarly: false },
  ],
  aaa: [
    { path: 'path/to/config/file', findup: false, exitEarly: false },
  ],
  bbb: [
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: '/root/path/to/file', findup: true, exitEarly: true },
  ],
  ccc: [[
    { path: '/root/path/to/file', findup: true, exitEarly: true },
    { path: 'path/to/file', findup: true, exitEarly: true },
   ],[
    { path: 'ddd/eee', findup: false, exitEarly: true },
    { path: 'fff/ggg', findup: false, exitEarly: true },
  ]],
  hhh: [
    { path: 'path/to/file', findup: false, exitEarly: false, userSpecKey: 'iii' }
  ]
}
--
(result)
configFiles = {
  rcFile: [
    { path: '/root/path/to/cwd/.apprc' },
  ],
  configs: [
    { path: '/root/path/to/homedir/config/.app' },
    { path: '/root/path/to/cwd/.config/app.conf' },
  ],
  aaa: [
    { path: '/root/path/to/config/file' },
  ],
  bbb: [
    { path: '/find/up/from/cwd/path/to/file' },
  ],
  ccc: [
    { path: '/root/path/to/file' },
    { path: '/root/path/to/fff/ggg' }
  ],
  hhh: [
    { path: '/root/path/to/file', userSpecKey: 'iii' },
  ]
}
```

phated replied:
I'm starting to like this. However, I think you can do away with `userSpecKey` and the final result should actually be objects that get loaded with `require` and their exported objects merged together.

phated replied:
Oh, and don't support an array as `path`.  Let me update your examples.

phated replied:
```
configFiles = {
  rcFile: '.apprc',
  configs: ['~/config/.app', './.config/app.conf'],
  aaa: { path: 'path/to/config/file' },
  bbb: [
    { path: 'path/to/file', findup: true, exitEarly: true  },
    { path: '/root/path/to/file', findup: true }
  ],
  ccc: [
    { path: '/root/path/to/file', findup: true, exitEarly: true },
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: 'ddd/eee', exitEarly: true },
    { path: 'fff/ggg' }
  ],
  hhh: { path: 'path/to/file' },
}
--
(expand, and process all in same way)
configFiles = {
  rcFile: [
    { path: '.apprc', findup: false, exitEarly: false },
  ],
  configs: [
    { path: '~/config/.app', findup: false, exitEarly: false },
    { path: './.config/app.conf', findup: false, exitEarly: false },
  ],
  aaa: [
    { path: 'path/to/config/file', findup: false, exitEarly: false }
  ],
  bbb: [
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: '/root/path/to/file', findup: true, exitEarly: true }
  ],
  ccc: [
    { path: '/root/path/to/file', findup: true, exitEarly: true },
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: 'ddd/eee', findup: false, exitEarly: true },
    { path: 'fff/ggg', findup: false, exitEarly: true }
  ],
  hhh: [
    { path: 'path/to/file', findup: false, exitEarly: false }
  ]
}
--
(result)
configFiles = {
  rcFile: { /* object from requiring .apprc */ }
  configs: { /* merged object from requiring /root/path/to/homedir/config/.app and /root/path/to/cwd/.config/app.conf */ },
  aaa: { /* object from requiring /root/path/to/config/file' */ },
  bbb: { /* object from requiring /find/up/from/cwd/path/to/file */ },
  ccc:  { /* object from requiring /root/path/to/file */ },
  hhh: { /* object from requiring /root/path/to/file */ }
}
```

Hope that helps

phated replied:
I'd also love to hear from @tkellen on this since it is straying from his original spec.  We will also be able to expand on `extensions` in the future but for now, we can rely on the extensions supported in the top-level Liftoff config.

sttk replied:
Ah, load and merge will be done automatically in this module. And rechoir selects appropriate loaders by getting extensions from searched paths. (I misunderstood about this.) That's good.

> I'd also love to hear from @tkellen on this

I also think it is good.

tkellen replied:
In my view, the only role for Liftoff here is locating the config files and ensuring the runtime is prepared to require them. All else should be handled by the consumer creating the tool.

tkellen replied:
Let me reframe... 

I'd support the mini dsl you're creating here @phated
```
 ccc: [
    { path: '/root/path/to/file', findup: true, exitEarly: true },
    { path: 'path/to/file', findup: true, exitEarly: true },
    { path: 'ddd/eee', findup: false, exitEarly: true },
    { path: 'fff/ggg', findup: false, exitEarly: true }
  ],
```

...as long as it is exposed via another module and upstreamed into Liftoff.

phated replied:
@tkellen that's fair, I forgot that gulp-cli is actually the thing requiring the final module, so it would make sense to leave it to consumers to require and merge as they see fit.  rechoir should already handle the runtime preparation if passed each filepath.  I also agree that it could be it's own module.  @sttk would you like me to create a repo in the js-cli org for you to implement this?  I guess we could drop the exitEarly if we are leaving the requiring up to the consumer (they would just require the first).

@tkellen the DSL was really just a refactor/simplification of your original spec, do you think it works for the use cases your spec was addressing?  It definitely solves the `.gulprc` use case I could think of.

sttk replied:
@phated Thanks, but will the module become redundant? If @tkellen implements this feature for liftoff, we will use it in gulp-cli. And I'm interested in @tkellen's DSL.

phated replied:
@sttk he meant that he wants it to be a separate module (similar to how `rechoir` is a separate module that Liftoff consumes).  Also, @tkellen won't be implementing anything, we are still using your work (which is excellent and thanks again).

sttk replied:
@phated If so, I'd like to create this. Please create a repository about this.

phated replied:
Sounds good. Would you mind continuing to work on all that was discussed above in this PR while we try to think of a good package name?

phated replied:
@sttk let me know when you have a chance to implement what was discussed above in this repo.  It's easiest to implement in this PR since you have all the tests, etc and we can refactor it to a separate repo afterwards.  Also let me know if I have to clarify anything.

phated replied:
Added an example spec to the original issue to fully flesh out my thoughts and show the differences between the original spec.

sttk replied:
@phated I will continue to implement this feature.

sttk replied:
@phated Ahhh, I'm sorry. I did misread 'when' and 'whether'.
> let me know when you have a chance to implement what was discussed above in this repo. 

I'll start to implement the new spec tomorrow and I'll finish it up next week sometime. 
Please clarify the spec about the questions I wrote in #8.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+    if (!opt || isString(opt) || isArray(opt)) {
+      opt = { searchPaths: opt };
+    } else {
+      opt = mapValues(opt, doNothing);
```
Is this used as a shallow clone?  Could we avoid mutations instead?

sttk replied:
OK. I'll do so.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+      opt = mapValues(opt, doNothing);
+    }
+
+    var srchPths, opeArr;
```
not sure what `opeArr` stands for, could you name it something more clear?  I'm fine with long variable names.  Might as well fully spell out `searchPaths` too.

sttk replied:
OK. I'll rename this variable.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+  ctx = normalizeContext(ctx);
+  return mapValues(opts || {}, searchWithOpt);
+
+  function searchWithOpt(opt, optname) {
```
does `optname` become annoying when opts are an array?  Should we reject arrays?

sttk replied:
I'll try to support cases that `opts` is a string or an array.

phated commented on lib/search_files.js:
```diff
@@ -0,0 +1,252 @@
+      opeArr = eachUnNullElem;
+    }
+
+    optname = normalizeHomedir(optname, ctx.homedir);
```
not sure I understand how optname is the path being passed to `normalizeHomdir`

sttk replied:
`optname` is used as a part of search paths. I think a case that `optname` contains `~`.

phated commented:
@sttk I'm sorry all these changes are coming in small chunks;  I'm still trying to decipher the entirety of the PR and the incremental improvements are helping a lot.

sttk commented:
@phated Thanks a lot. I appreciate your review in detail and I'm pleased that the codes are gradually getting better. I'll be glad to accept and modify what you pointed out.

sttk commented:
I've finish implementing and testing new spec.
(I renamed the module file name from 'search_files.js' to 'multi_configs.js'.)

sttk commented:
Sorry. I found a bug and fixed it.

phated commented:
@sttk I noticed that you implemented use of `extension` property in the configFiles objects but you didn't use `rechoir` to prepare the environment (e.g. `.coffee` would load `coffee-script/register`)

phated commented on lib/multi_configs.js:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  return mapValues(options, function(opt, name) {
+    return firstUnNull(expandEachProp(opt, name), findWithExpandedPaths);
```
use `lodash.find` instead of a custom `firstUnNull` method - https://lodash.com/docs#find

sttk replied:
@phated The usage of `lodash.find` is similar to `lodash.filter` and it doesn't return a result by argument function. So if it is substituted for `firstUnNull`, the code would be as follows:

```js
  return (lodash.find(expandEachProp(opt, name), function(expandedPath) {
    return (expandedPath.searchedPath = findWithExpandedPaths(expandedPath)) != null;
  }) || {}).searchedPath;
```

Is this code along your thoughts?

phated replied:
Oh yeah, I was thinking about `lodash.find` incorrectly.  I also forgot that `findWithExpandedPaths` needed to return a different path than `expanded.path` if `findUp` is true.

phated replied:
No need to use `lodash.find` then.

sttk replied:
I got it.

phated commented on lib/multi_configs.js:
```diff
@@ -0,0 +1,129 @@
+  });
+
+
+  function expandEachProp(opt, name) {
```
could the amount of code be reduced by combining all the expand/normalize methods - `expandEachProp` calls `createExpandedPathArray` which calls `normalizeExpandedPath` but I don't think they are called anywhere else.  It might be simpler to combine them and just add logic based on the state of the `opt` (object, string, etc)

sttk replied:
I'll combine those three functions into one.

phated commented on lib/multi_configs.js:
```diff
@@ -0,0 +1,129 @@
+  return null;
+}
+
+function findup(filepath, basedir) {
```
can you use `findup-sync` for this now?

sttk replied:
If we don't have to support glob, `findup-sync` will not be needed.

phated commented on lib/multi_configs.js:
```diff
@@ -0,0 +1,129 @@
+  return def || [''];
+}
+
+function find(filepath) {
```
use https://github.com/sindresorhus/path-exists#pathexistssyncpath instead of a custom implementation

sttk replied:
Since it isn't needed to check whether a file or not, this `find` function will become to execute only `fs.existsSync`.

phated replied:
Why do you need to check it is a file?  We should be able to resolve directories too because `require('somedir')` will still load, given their is an `index.js` file or a `package.json` describing the main file.

We should only care that something exists; whether it is a file, directory, symlink should be irrelevant to us.

sttk replied:
I did misunderstand. The reasons of misunderstanding were: 

* Since extensions are specified basically (liftoff use `.js, .json` by default), users need to specify `extensions` property empty for searching a directory. (But I thought it is not intuitive.)
* A directory may hit wrongly though a user want a file which has same name.
* Even if a directory is searched, `index.js` or `package.json` may not exist in the directory.

But these would be what users should care in accordance with their cases.

phated commented on lib/multi_configs.js:
```diff
@@ -0,0 +1,129 @@
+function find(filepath) {
+  try {
+    var stat = fs.statSync(filepath);
+    if (stat.isFile()) {
```
why do you need to check if it is a file?  node require has special semantics around requiring a directory

phated commented:
@sttk thought of a name for the module - https://github.com/js-cli/fined

jonschlinkert commented:
nice, that's an awesome name

sttk commented:
@phated I also think that's a fine name!
I'll modify what you pointed out above, so please wait.

tkellen commented:
@sttk I haven't had a chance to review this closely, but when you implement this stuff over at `fined` please make sure that the new module has no semantics for trying to find multiple files. Its sole responsibility should be to accept a configuration for a single file and locate it. Liftoff should be responsible for composing multiple calls to `fined`.

sttk commented:
@tkellen I got it. I misunderstood that the searched file should have been a file and I had forgotten to check a file because of specifying extensions. This misunderstanding was one of the reasons that I didn't adopt `findup-sync`.

And another reason is about supporting glob.

@tkellen @phated Should `fined` support glob? Or should `fined` use `findup-sync` with avoiding glob by using `is-glob`?

jonschlinkert commented:
> should fined support glob? Or should fined use findup-sync with avoiding glob by using is-glob?

whatever approach you use, anything you can do to reduce calls to fs will make it faster. I can't think of anything that would slow the code down more than hitting the file system. just my 2c

sttk commented:
@phated 

>@sttk I noticed that you implemented use of extension property in the configFiles objects but you didn't use rechoir to prepare the environment (e.g. .coffee would load coffee-script/register)

Considering the case that `extension` property is specified an interpret-like configuration object, I think it's better that the result of each searched path should be an object which consists of the path and the hitting extension info than only the path string.

`fined` will not load searched files automatically, but users will be able to load it easily by using the extension info.

For example:

```
(input)
configFiles = {
  app: {
    path: '.',
    extensions: ['.json', '.js']
  },
  appfile: {
    path: '.',
    extensions: require('intepret').jsVariants
  }
}

(result)
env.configFiles = {
  app: {
    path: '/path/to/app.js',
    extension: ['.js']
  },
  appfile: {
    path: '/path/to/appfile.coffee',
    extension: { '.coffee': ['coffee-script/register', 'coffee-script'] },
                            // this value is got by `extensions['.coffee']`
  }
};
```

sttk commented:
I've finished modifying the code.

I implemented tentatively that the result of each searched path is an object as I wrote above.

tkellen commented on lib/fined.js:
```diff
@@ -0,0 +1,141 @@
+const parsePath = require('parse-filepath');
+
+
+function search(options, extensions, cwd) {
```
The purpose of this function should be to find a single config file. The object iteration you've embedded here belongs outside of the `fined` library, or at the very least in a higher order method that calls the `search` primitive repeatedly.

That is to say you should not be able to do this:
```js
search({
  app: {
    path: path.resolve(cwd, 'test/fixtures/fined/app.js'),
    extension: '.js',
  },
  package: {
    path: path.resolve(cwd, 'package.json'),
    extension: '.json'
  }
}, ....);
```

This function should only take the value side of a single configFiles object.
```js
search({
  path: path.resolve(cwd, 'test/fixtures/fined/app.js'),
  extension: '.js',
}, ...)

search({
  path: path.resolve(cwd, 'package.json'),
  extension: '.json'
}, ...)
```

tkellen replied:
Put another way, there is no reason for the search method to know about the `app` and `package` keys. The purpose of the `fined` library is to find config files, not map exactly to the Liftoff API.

tkellen replied:
Put yet *another* way, you should never embed iteration into a primitive. Make your primitive act on a single atomic unit. Having done that you can use any control flow you like to call the primitive multiple times.

In some sense this is what you've done:
```js
function doubler(input) {
  if (Array.isArray(input)) { 
    return input.map(x => x * 2);
  }
  return input * 2;
}
doubler(1);
doubler([1,2,3]);
```

...when this is much easier to reason about / unit test / etc:
```
function doubler(input) {
  return input * 2;
}
doubler(1);
[1,2,3].map(doubler);
```

tkellen replied:
PS: this is really great work so far, thank you so much!

sttk replied:
@tkellen I've understood now that a single file you say is really a single file to be found by `fined` at a time. I misunderstood that it was a single file by each properties of `configFiles` and `fined` found files for all properties at a time. I'm sorry.

tkellen replied:
No need to be sorry @sttk, I am **so** appreciative of the work you are doing, thank you again!

sttk replied:
I'll try to modify `fined` (`search` function) to process a single property and find a single file at a time.

sttk commented:
@tkellen I've understood now that a single file you say is really a single file to be found by `fined` at a time. I misunderstood that it meant a single file **by each properties** of `configFiles` and `fined` found files for all properties at a time. I'm sorry.

sttk commented:
I've modified the code as follows:

* `fined` processes a single property of `configFiles` and finds a single file at a time.
* About result form, `fined` returns an object which consists of a path string and an extension string/object. On the other hand `liftoff` sets only a path string to each property of `env.configFiles`.
* `liftoff` uses `rechoir` also for extension objects of files found by `fined`. (tentatively)

phated commented:
@sttk hoping to review the changes soon.  Been busy, sorry.

phated commented on index.js:
```diff
@@ -122,21 +125,37 @@ Liftoff.prototype.buildEnvironment = function (opts) {
+
+  var configFiles = {};
+  if (isPlainObject(this.configFiles)) {
+    var fined = new Fined(cwd, this.extensions);
```
Initial thoughts (without thoroughly reviewing anything) - can you avoid the constructor pattern?  I'd prefer for `fined` to just be a function that can be called instead of using `new` then calling a `.find` method on the instance.

sttk replied:
This constructor do some processes for the default extensions. Without this constructor, these processes have to be done for every `configFiles` property with no `extensions` property.

phated replied:
@sttk what about using memoization or caching instead?

sttk replied:
@phated I think memoization would not be suitable in this case, because what should be cached is only results of default `extensions`. And caching seems not to be good when passed same `extensions` object but changed its properties. 

What I worried about was mainly the cost of `Object.keys` for an object which has a lot of properties like `js-interpret.extensions`, but it is not to be necessary to worry about (it takes about 0.1ms on my environment.)

phated replied:
Cool, since it isn't an issue, implement it with a pure function instead of the constructor pattern and if we need to optimize it later, we can dig into that separately.

tkellen replied:
:+1: to using pure functions

sttk replied:
@phated @tkellen I got it. I'll modify to using a pure function.

I'm considering the interface of this function. The parameters which `fined` has to receive are:

   * **pathObj** : a path object or a path string, and if it is an object the properties are:
       * a path string
       * a find-up flag
       * extensions, which can be a string, an array or an object.
   * **name** : a base name of searched file. (`configFiles`'s property key in `liftoff`)
   * **cwd** : a default base directory path.
   * **exts** : a default extensions, which can be a string, an array or an object.

And about the interface,

1. the most simple interface is:

   ```js
   var pathObj = { path: (mandatory), findUp: (optional) , extensions: (optional) };
   var file = fined(pathObj, name, cwd, exts); // `name`, `cwd`, `exts` are optional.
   ```

   But I think it would be confusing about these order for consumers.

2. the interface now I'm thinking are:

   ```js
   var pathObj = { path: (optional), findUp: (optional), extensions: (optional), cwd: (optional), name: (optional) };
   var defaultObj =  { path: (optional), findUp: (optional), extensions: (optional), cwd: (optional), name: (optional) };

   var file = fined(pathObj, defaultObj)
  ```

   `pathObj` and `defaultObj` can have same properties and lacking properties in `pathObj` are complemented with properties in `defaultObj` (by using `lodash.defaults`).

What do you think?

tkellen replied:
Thanks again for all the work you're doing here @sttk, we're getting really close! I am very sorry but I don't have time to dig into this right now :(. @phated can you chime in?

tkellen commented on lib/fined.js:
```diff
@@ -0,0 +1,151 @@
+  return extArr.map(createExpandedPath.bind(this));
+
+
+  function createExpandedPath(ext) {
```
I would avoid creating functions inside of functions... just define this outside the scope of the main function body and pass in the values you need. This will make it a zillion times easier to test, as well.

sttk replied:
I got it. I'll define this function outside the main function.

tkellen commented on lib/fined.js:
```diff
@@ -0,0 +1,151 @@
+  return { path: found, extension: expanded.extension };
+
+
+  function findFile(filepath) {
```
I would lift this out of this function--it doesn't rely on any state and doesn't need to be created dynamically with every invocation of `findWithExpandedPath`

sttk replied:
I'll modify so.

tkellen commented on lib/fined.js:
```diff
@@ -0,0 +1,151 @@
+    return null;
+  }
+
+  function findUpFile(filepath, basedir) {
```
same for this function

tkellen replied:
getting really close here!

sttk commented:
I've remake `fined` as a pure function.

tkellen commented on lib/fined.js:
```diff
@@ -0,0 +1,149 @@
+    return [exts];
+  }
+
+  if (isArray(exts)) {
```
124-134 could be
```
if (isArray(exts)) {
  exts = exts.filter(isString);
}
if (isPlainObject(exts)) {
  exts = Object.keys(exts);
}

return exts.length ? : ext : [''];
``` 

right?

sttk replied:
Yes, you're right.

sttk replied:
@tkellen Sorry. I take back my comment. Your code above can cause error when `exts` is illegal type other than a string, an array and a plain object. (Though no one would specify such a value).

tkellen commented:
Hot damn, I think we might be there @sttk. Let's give @phated a chance to do a final pass and then decide how to proceed? I think perhaps we should move `fined` to its own module and upstream it here vs inlining it?

tkellen commented:
@sttk, are you on twitter?

sttk commented:
No, I don't have an account on twitter.
And I'm sorry that I'm going to work now. (It's morning in Japan now.)

phated commented on lib/fined.js:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  var expanded = [];
+  for (var i = 0, n = extArr.length; i < n; i++) {
```
use `extArr.map()` here instead of a `for` loop with `expanded.push`

phated commented on lib/fined.js:
```diff
@@ -0,0 +1,149 @@
+
+function fined(pathObj, defaultObj) {
+  var expandedArr = expandPaths(pathObj, defaultObj);
+  for (var i = 0, n = expandedArr.length; i < n; i++) {
```
Is this using a `for` loop to allow exiting early on the first `found`?

sttk replied:
@phated Yes. This part was a function `firstUnNullElem` but I think it may be misunderstandable and the function would be slightly slower than `for` loop block. Is it better to use a function here?

phated commented:
I had a couple comments but in general, this looks amazing.  Once those few things are clarified and adjusted, I think `fined` can be submitted as a PR to the repository I created.  I'll try to get it scaffolded tomorrow.

phated commented:
Thanks for all the hard work @sttk!!!

sttk commented:
@phated I've modified a `for` loop to array.map.

phated commented:
@sttk excellent.  I scaffolded the `fined` repository.  Can you send a PR to that repo with the implementation you came up with here (tests too)?

sttk commented:
@phated I've sended a PR to the `fined` repository. Please change it as you want, especially README file because I have no confidence in my English.

jonschlinkert commented:
fwiw @sttk your english is great.

sttk commented:
@jonschlinkert Thanks, but I write comments with a dictionary. English conversation is more difficult than programming for me.

jonschlinkert commented:
> English conversation is more difficult than programming for me.

Same for me lol, and english is my native language

phated commented:
After all the amazing work by @sttk, `fined` 1.0.0 has just been published.  This PR can now be updated to use that module as a dependency.

phated commented:
Oh, please rebase it also so I can merge cleanly.

sttk commented:
Sorry. I found a bug of `fined` (PR [#2](https://github.com/js-cli/fined/pull/2)).

@phated Now I'm re-creating this from master branch. And I'll push this after reflecting bug-fixed `fined`. Please review the PR [#2](https://github.com/js-cli/fined/pull/2).

sttk commented:
I've finished modifying!

phated commented on index.js:
```diff
@@ -121,22 +124,21 @@ Liftoff.prototype.buildEnvironment = function (opts) {
-          this.emit('requireFail', attempt.moduleName, attempt.error);
-        } else {
-          this.emit('require', attempt.moduleName, attempt.module);
+  registerLoader(this.extensions, configPath, cwd, this);
```
I'd recommend moving the `this` from the last argument to the first.  Maybe this function signature is getting too complex and should be changed to an object with each argument as a property.  @tkellen thoughts?

phated commented on index.js:
```diff
@@ -121,22 +124,21 @@ Liftoff.prototype.buildEnvironment = function (opts) {
+
+  var configFiles = {};
+  if (isPlainObject(this.configFiles)) {
+    var exts = this.extensions, notfound = { path: null }, eventEmitter = this;
```
`var exts = this.extension` can be defined higher in the function and used with the first `registerLoader` call also.

phated replied:
same with `eventEmitter`

phated replied:
Every variable declaration should be it's own line with `var ...`

sttk replied:
I'll modify so.

phated commented on index.js:
```diff
@@ -121,22 +124,21 @@ Liftoff.prototype.buildEnvironment = function (opts) {
+    var exts = this.extensions, notfound = { path: null }, eventEmitter = this;
+    configFiles = mapValues(this.configFiles, function(prop, name) {
+      var defaultObj = { name: name, cwd: cwd, extensions: exts };
+      return mapValues(prop, function(pathObj) {
```
Is 2-level nested objects something we actually want to support?  /cc @tkellen (example usage https://github.com/js-cli/js-liftoff/pull/67/files#diff-910eb6f57886ca16c136101fb1699231R326)

phated commented on index.js:
```diff
@@ -121,22 +124,21 @@ Liftoff.prototype.buildEnvironment = function (opts) {
+      var defaultObj = { name: name, cwd: cwd, extensions: exts };
+      return mapValues(prop, function(pathObj) {
+        var found = fined(pathObj, defaultObj) || notfound;
+        if (isPlainObject(found.extension)) {
```
doesn't `fined` allow this to be an array in some cases?

sttk replied:
`found.extension` is either a string or a plain object which contains only an extension of a file which is found first.

phated commented on index.js:
```diff
@@ -146,10 +148,30 @@ Liftoff.prototype.buildEnvironment = function (opts) {
 
+// use rechoir to autoload any required modules
+function registerLoader(extensions, configPath, cwd, eventEmitter) {
+  if (configPath) {
```
instead of this nested `if` statement, check `if (!configPath)` and return early

sttk replied:
I'll modify so.

phated commented on package.json:
```diff
@@ -37,7 +37,10 @@
   "dependencies": {
     "extend": "^3.0.0",
     "findup-sync": "^0.4.1",
```
@sttk are we going to be able to drop the `findup-sync` usage and replace with `fined`?

phated replied:
If so, that will probably be a separate PR.

phated replied:
I had to roll this back because of a bug.  Please rebase to get that change.

sttk replied:
I got it. I'll reflect changes of master to this.

phated commented:
@sttk except for my few notes and questions, this looks great!

sttk commented:
@phated Thank you for pointing out. I rebased and reflected changes of master. In addition, I modified something that you pointed out above and I could do immediately.

phated commented:
@sttk sorry for the delayed review.  This looks amazing.  Can you believe that you reduced the PR to 300 lines of code (most of it being tests)?  I'd love to get @tkellen and/or @jonschlinkert to review before I click the merge button.

tkellen commented on index.js:
```diff
@@ -146,10 +150,31 @@ Liftoff.prototype.buildEnvironment = function (opts) {
 };
 
+// use rechoir to autoload any required modules
+function registerLoader(eventEmitter, extensions, configPath, cwd) {
```
this should be in `lib/register_loader.js`. ideally it'd be unit tested, too!

tkellen commented:
Incredible work @sttk. If you have the appetite to address my final concern in this PR, awesome. If you'd rather do that as a separate PR after this is landed and published, I'm down with that too.

tkellen commented:
PS: Thank you!!

sttk commented:
@tkellen Thank you for reviewing. I'll address what you pointed out in this PR.

sttk commented:
I've separated `registerLoader` and unit tested it.

phated commented:
@sttk this looks great!  Can you rebase one last time?  I merged an update to the findup-sync module and it conflicts with some of your changes.  Once you rebase, I'll merge this in the next couple of days (I'd like to make sure the findup-sync change isn't causing problems before bumping again).

sttk commented:
@phated I've rebased.

phated commented:
@tkellen I'm looking to ship this today, you good with me doing that?

tkellen commented:
:shipit:

phated commented:
@sttk thank you so much for your hard work on this!  I'm going to be adding docs and then publishing this feature.

sttk commented:
Thank you very much!!

phated commented:
This was just published as 2.3.0.  Once again, amazing work.
